### PR TITLE
Add box storage support for transaction signing

### PR DESCRIPTION
**Pull request by k13n**
This adds box storage support for creating and signing transactions and addresses issue #191.

### CODE REVIEW

**Comment by manuelmauro** [comment]
Thanks a lot for this, @k13n! It looks very good to me, I'll play a bit more with it and come back to you by tomorrow. 👍

File: algonaut_transaction/src/api_model.rs [hunk 1]
```diff
@@ -567,6 +593,99 @@ fn int_to_application_call_on_complete(
     }
 }
 
+fn api_box_references_from_box_references(
```
Comment (manuelmauro) [inline algonaut_transaction/src/api_model.rs:596]: We could encode these functions as type conversions, here a gist with the main idea (could be further improved) https://gist.github.com/manuelmauro/0da03cde317596e064e2017c891edf81

File: algonaut_transaction/src/api_model.rs [hunk 2]
```diff
@@ -614,4 +733,57 @@ mod tests {
...
+                name: box_name.clone(),
+            },
+        ];
+
+        let api_boxes = api_box_references_from_box_references(
+            Some(&boxes),
+            Some(&foreign_apps),
+            Some(current_app_id),
+        )
+        .expect("api box references can be created");
```
Comment (manuelmauro) [inline algonaut_transaction/src/api_model.rs:762]: nit: Rust's recommended message style for `expect` would be something like "api box should be created" (ref: https://doc.rust-lang.org/std/option/enum.Option.html#recommended-message-style)

**Comment by k13n** [comment]
Thanks @manuelmauro! I just updated the PR based on your gist.

File: algonaut_transaction/src/api_model.rs (hunk 1 above)
Comment (manuelmauro) [reply algonaut_transaction/src/api_model.rs]: :) I was not super happy with `BoxesInfo` and `ApiBoxesInfo` and I wonder if just moving their code into `impl TryFrom<Transaction> for ApiTransaction`and `impl TryFrom<ApiTransaction> for Transaction` would be better (probably no, since we use the conversion also in the testing and we would need to duplicate that).
Also, the trait `AppCallMetadata` sounds like a good idea and it probably deserves a place somewhere else, but I'd leave it there for now and see if it pops up again in the future.

In any case, I am happy to merge the PR whenever you want. I'd like to fix the SDK testing CI in the near future to have a better coverage, but for now the test you wrote look good to me!

Comment (k13n) [reply algonaut_transaction/src/api_model.rs]: Would it make sense to have `fn from_box_reference(box_, current_app_id, foreign_apps)` in `impl ApiBoxReference` and a counterpart in `impl BoxReference` to convert from an `ApiBoxReference`? Probably not, since they're located in different modules and I'm not sure how rusty that would be.

I think it'd be okay to merge & use what we have now. What I like is that `BoxesInfo` and `ApiBoxesInfo` is used only internally and doesn't leak to the caller. Therefore, it can be refactored in the future if a better idea pops up.

Comment (manuelmauro) [reply algonaut_transaction/src/api_model.rs]: Sorry, I missed this for some reason! Gonna merge it then!
